test/others/criu-ns/run.py

```diff
@@ -0,0 +1,181 @@
+#!/usr/bin/env python
+
+import os
+import pty
+import shutil
+import subprocess
+import sys
+import time
+
+
+CRIU_BIN = "../../../criu/criu"
+CRIU_NS = "../../../scripts/criu-ns"
+IMG_DIR = "dumpdir"
+PIDFILE = "pidfile"
+
+
+def check_dumpdir():
+    if os.path.isdir(IMG_DIR):
+        shutil.rmtree(IMG_DIR)
+    os.mkdir(IMG_DIR, 0o755)
+
+
+def create_pty():
+    fd_m, fd_s = pty.openpty()
+    return (os.fdopen(fd_m, "wb"), os.fdopen(fd_s, "wb"))
+
+
+def create_isolated_dumpee():
+    with open("running", "w") as file:
+        file.name
+    fd_m, fd_s = create_pty()
+    pid = os.fork()
+    fd_m.close()
+    if pid == 0:
+        os.setsid()
+        os.dup2(fd_s.fileno(), 0)
+        os.dup2(fd_s.fileno(), 1)
+        os.dup2(fd_s.fileno(), 2)
+        while True:
+            if not os.access("running", os.F_OK):
+                sys.exit(0)
+            time.sleep(1)
+    return (pid, fd_s)
+
+
+def criu_ns_dump(pid, shell_job=False):
+    if shell_job:
+        cmd = [CRIU_NS, "dump", "-D", "dumpdir", "-v4",
+               "--shell-job", "-t", str(pid), "--criu-binary",
+               CRIU_BIN]
+    else:
+        cmd = [CRIU_NS, "dump", "-D", "dumpdir", "-v4",
+               "-t", str(pid), "--criu-binary", CRIU_BIN]
+    ret = subprocess.Popen(cmd).wait()
+    return ret
+
+
+def criu_ns_restore(shell_job=False, restore_detached=False):
+    if shell_job:
+        cmd = [CRIU_NS, "restore", "-D", "dumpdir", "-v4",
+               "--shell-job", "--criu-binary", CRIU_BIN]
+    else:
+        cmd = [CRIU_NS, "restore", "-D", "dumpdir", "-v4",
+               "--criu-binary", CRIU_BIN]
+    if restore_detached:
+        cmd = [CRIU_NS, "restore", "-D", "dumpdir", "-v4",
+               "--restore-detached", "--pidfile", PIDFILE,
+               "--criu-binary", CRIU_BIN]
+    ret = subprocess.Popen(cmd).wait()
+    return ret
+
+
+def test_dump_and_restore_with_shell_job():
+    """Test criu-ns dump and restore with --shell-job option"""
+    check_dumpdir()
+    os.setsid()
+
+    with open("running", "w") as file:
+        file.name
```

Comment:
Why have you changed it from `pass` to `file.name`?

Reply:
Ah I see it's linter. Sorry

Reply:
What about using `from pathlib import Path`?

```Python
Path('running').touch()`
```

Reply:
`Path('running').touch()` works great, thanks @rst0git

Reply:
pathlib does not seem to be present in centos 7, that is why now we have test on centos 7 failing

Reply:
should I bring back the following?
```
with open("running", "w") as file:
        file.name
```

Reply:
cc @rst0git

Reply:
@warusadura We can install `pathlib` in the CentOS 7 environment. Could you try the following patch?
```diff
diff --git a/.cirrus.yml b/.cirrus.yml
index bd4799fd0..80f3296fc 100644
--- a/.cirrus.yml
+++ b/.cirrus.yml
@@ -134,11 +134,11 @@ task:
     memory: 8G
 
   setup_script: |
-    # EPEL is needed for python2-future, python2-junit_xml, python-flake8 and libbsd-devel.
+    # EPEL is needed for python2-future, python2-junit_xml, python-pathlib, python-flake8 and libbsd-devel.
     # Do not fail if latest epel repository definition is already installed
     yum install -y https://dl.fedoraproject.org/pub/epel/epel-release-latest-7.noarch.rpm || :
     ln -sf /usr/include/google/protobuf/descriptor.proto images/google/protobuf/descriptor.proto
-    yum install -y findutils gcc git gnutls-devel iproute iptables libaio-devel libasan libcap-devel libnet-devel libnl3-devel libbsd-devel make procps-ng protobuf-c-devel protobuf-devel protobuf-python python python-flake8 python-ipaddress python2-future python2-junit_xml python-yaml python-six sudo tar which e2fsprogs python2-pip rubygem-asciidoctor libselinux-devel
+    yum install -y findutils gcc git gnutls-devel iproute iptables libaio-devel libasan libcap-devel libnet-devel libnl3-devel libbsd-devel make procps-ng protobuf-c-devel protobuf-devel protobuf-python python python-flake8 python-ipaddress python2-future python2-junit_xml python-yaml python-six python-pathlib sudo tar which e2fsprogs python2-pip rubygem-asciidoctor libselinux-devel
     # Even with selinux in permissive mode the selinux tests will be executed
     # The Cirrus CI user runs as a service from selinux point of view and is
     # much more restricted than a normal shell (system_u:system_r:unconfined_service_t:s0)

```

Reply:
Thanks @rst0git. Is it alright to introduce this change in this PR?

Reply:
> Is it alright to introduce this change in this PR?

Yes.

Note that the tests in CentOS 7  are using Python 2 and you might need to change the shebang line of `criu-ns` for the tests to work.

Reply:
understood, thanks!

Reply:
@rst0git `run.py` test script contains function calls such as `print()`https://github.com/checkpoint-restore/criu/blob/34a95a83cb7f4f74ef9d743df135c284c82f8228/test/others/criu-ns/run.py#L74 Since the CentOS 7 CI job is based on Python 2, I think the tests will fail :)

Reply:
> Since the CentOS 7 CI job is based on Python 2, I think the tests will fail :)

Assuming that the shebang like is changed as follows, the `criu-ns` tests pass when I run them locally with CentOS 7 and Python 2.

```
sed -i 's/python3/python/' scripts/criu-ns
```

Reply:
understood, thanks @rst0git

Reply:
CentOS 7 CI job now fails with the following error,
`AttributeError: 'module' object has no attribute 'set_blocking'`

`os.set_blocking` introduced in Python 3.5

I think it make sense to bring back the following,
```
with open("running", "w") as file:
        file.name # to avoid linter warnings
```
@rst0git

Reply:
sorry, said the wrong thing :) 
Is there a way to remove the CentOS 7 CI tests from executing against `criu-ns` tests?

Reply:
@warusadura Could you replace `os.set_blocking` with something like the following?
```Python
import os
import fcntl

def set_blocking(fd, blocking):
    """Implement os.set_blocking() for compatibility with Python versions earlier than 3.5"""
    flags = fcntl.fcntl(fd, fcntl.F_GETFL)
    
    if blocking:
        flags &= ~os.O_NONBLOCK
    else:
        flags |= os.O_NONBLOCK
    
    fcntl.fcntl(fd, fcntl.F_SETFL, flags)
```

Reply:
Done, thanks @rst0git. 

I'm just curious, why do we need to test new changes against older versions of Python such Python 2?